### PR TITLE
[protocolv2] swap ReadStream's waitForClose promise with callback style onClose

### PR DESCRIPTION
**Pull request by masad-frost**
## Why

We discussed this in some of the earlier PRs for ReadStream. After using it, I decided that it's a lot cleaner to do a callback and wrap with a promise as needed instead of using `.then` as a callback when needed

## What changed

Title says it all

### CODE REVIEW

File: router/streams.ts [hunk 1]
```diff
@@ -136,13 +137,9 @@ export class ReadStreamImpl<T> implements ReadStream<T> {
...
   /**
-   * Used to for `waitForClose` and `requestClose`.
+   * A list of listeners that will be called when the stream is closed.
    */
-  private closePromise: Promise<undefined>;
-  /**
-   * Resolves closePromise
-   */
-  private resolveClosePromise: () => void = () => undefined;
+  private onCloseListeners: Set<() => void>;
```
Comment (Monkatraz) [inline router/streams.ts:142]: There is a utility exported from `@replit/observable` called `ListenerList` that would be helpful here. You can use it like this:
```ts
const list = new ListenerList()
const dispose = list.add(() => {})
list.notify()
```
or, if you want to notify with arguments:
```ts
const list = new ListenerList<(msg: string, int: number) => void>()
const dispose = list.add((msg, int) => {})
list.notify('foo', 1)
// notify inherits the type parameter's arguments
```

**Review by Monkatraz — APPROVED** [review]
lgtm, just one comment about maybe using `ListenerList`

File: router/streams.ts (hunk 1 above)
Comment (masad-frost) [reply router/streams.ts:142]: Looks like we haven't published it

Comment (masad-frost) [reply router/streams.ts:142]: this seems simple enough, it's not really worth a utility imo

Comment (masad-frost) [reply router/streams.ts:142]: if we end up with a lot of listeners, we can swap them out.

Comment (Monkatraz) [reply router/streams.ts:142]: oh right I totally forgot that this is in river lmao

**Pull request merged**